### ISSUE TITLE
Brave anti-fingerprint fails against fingerprint.js

### ISSUE THREAD

**Issue by devingoodman** [brave/brave-browser#14031]
1. Go to https://fingerprintjs.com/demo/
   2. Open new tabs, new profiles. Try again. Make sure your shields are on and all turned on to strict/aggressive.
   3. Delete your cookies, history, etc. and try again.

## Actual result:

Screenshot attached

## Reproduces how often: 
Easily reproduced.

## Desktop Brave version:
Brave	1.19.92 Chromium: 88.0.4324.152 (Official Build) (x86_64)
Revision	6579930fc53b4dc589c042bec9d0a3778326974d-refs/branch-heads/4324@{#2106}
OS	macOS Version 11.2 (Build 20D64)
JavaScript	V8 8.8.278.15
User Agent	Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.152 Safari/537.36
Command Line	/Applications/Brave Browser.app/Contents/MacOS/Brave Browser --enable-dom-distiller --disable-domain-reliability --no-pings --extension-content-verification=enforce_strict --extensions-install-verification=enforce --origin-trial-public-key=bYUKPJoPnCxeNvu72j4EmPuK7tr1PAC7SHh8ld9Mw3E=,fMS4mpO6buLQ/QMd+zJmxzty/VQ6B1EUZqoCU04zoRU= --sync-url=https://sync-v2.brave.com/v2 --lso-url=https://no-thanks.invalid --variations-server-url=https://variations.brave.com/seed --enable-features=AutoupgradeMixedContent,WebUIDarkMode,LegacyTLSEnforced,PrefetchPrivacyChanges,PasswordImport,ReducedReferrerGranularity --disable-features=AutofillEnableAccountWalletStorage,TextFragmentAnchor,NotificationTriggers,WebOTP,PrivacySettingsRedesign,PasswordCheck,NetworkTimeServiceQuerying,TabHoverCards,AutofillServerCommunication,IdleDetection,SignedExchangeSubresourcePrefetch,SafeBrowsingEnhancedProtection --flag-switches-begin --load-media-router-component-extension=1 --enable-features=AutoupgradeMixedContent,WebUIDarkMode,LegacyTLSEnforced,PrefetchPrivacyChanges,PasswordImport,ReducedReferrerGranularity,DnsOverHttps,LazyImageLoading:automatic-lazy-load-images-enabled/true/restrict-lazy-load-images-to-data-saver-only/false,ParallelDownloading,WebRtcUseMinMaxVEADimensions --flag-switches-end
Executable Path	/Applications/Brave Browser.app/Contents/MacOS/Brave Browser
Profile Path	/Users/redacted/Library/Application Support/BraveSoftware/Brave-Browser/Default
Variations	2729b628-70ea8f25

## Version/Channel Information:
Stable

## Other Additional Information:

- Does the issue resolve itself when disabling Brave Shields? No
- Does the issue resolve itself when disabling Brave Rewards? No
- Is the issue reproducible on the latest version of Chrome? Yes
<img width="634" alt="Screen Shot 2021-02-07 at 11 54 46 PM" src="https://user-images.githubusercontent.com/7727156/107191629-4993b480-69a1-11eb-9c25-b876cedee970.png">
<img width="634" alt="Screen Shot 2021-02-07 at 11 54 46 PM" src="https://user-images.githubusercontent.com/7727156/107191661-5a442a80-69a1-11eb-897f-30115d1cc410.png">

**Comment by Smagold** [comment]
I can confirm. The fingerprint ID remains the same even if you change your IP address. With Brave's "strict protection" mode on there is still some unique data they are able to record.
Also, the bizarre thing is – they somehow distinguish Brave from other browsers. I have a fake User Agent set in both `navigator` & the HTTP header, and yet they are able to tell that I'm on Brave.

**Comment by pes10k** [comment]
Thank you for the report @Smagold and @devingoodman , i will take a look and report back!

**Comment by TheCultLeader666** [comment]
I have this problem as well, although this seems to affect all browsers, even Tor (although it did change after I reset my identity)

**Comment by TheCultLeader666** [comment]
It seems (through my own testing) that the website additionally uses the cache to identify users (or maybe favicons? From my understanding Brave now includes the favicon cache when clearing browser cache), as whenever I clear my cache the ID resets. 

> Also, the bizarre thing is – they somehow distinguish Brave from other browsers.

My best guess for how it would identify Brave as the browser is that it might be able to pick out Brave Shields because of how it tries to block fingerprinting, but I am nowhere near qualified enough to come to any concrete conclusions.

**Comment by bsclifton** [comment]
@Smagold we do have a custom property available called `navigator.brave.isBrave` which is present to help partners identify this. @pes10k helped define the approach taken and can share more details (I did the implementation on Desktop/Android). We don't have a way to turn this off

**Comment by Smagold** [comment]
> we do have a custom property available called `navigator.brave.isBrave` which is present to help partners identify this.

@bsclifton I have a script that, among other things, does this on every page:
```javascript
if (navigator.brave) {
  delete Navigator.prototype.brave;
}
```
And yet, FingerprintJS demo page is able to tell that I'm on Brave: `Brave on Windows (10)`.

My User Agent, Platform, and some other navigator properties are faked and FingerprintJS seems to trust the faked values no problem. That makes me think that it's possible they detect Brave based on some other non-navigator properties that are leaking out.

There isn't as many Brave users as we all wish there was, and seeing a script being able to tell us apart in the crowd of other browsers makes it possible to target us with some more aggressive & precise fingerprinting tactics. I think it's an issue worth investigating.

**Comment by pes10k** [comment]
Hmm, i think the fact that the site caches responses in storage, and uses that to look like its fingerprinting you is making folks think this thing is more clever than it is :P

I just tested now, once in my main profile, and once in a tor tab (to get a new profile, new window and a new IP), and, as expected, the site wasn't able to link the two 

Are folks seeing otherwise if they repeat the above?

I dug into this a while ago, and at least then, the site was doing the following:

1. looking at storage to "fingerprint" you (i.e. not fingerprinting at all)
2. using IP + useragent (again, not fingerprinting)
3. if still no match, use fingerprint2.js signals, and send them server side.  And for brave users, drop the inputs we randomize (i.e., a good thing for brave users).  If there is a match, report a match, otherwise they give up

Anyway, TL;DR; the site is mostly smoke and mirrors trying to sell a commercial service.  They're one of the sites we use when deciding what to randomize next for fingerprinting protections, but they're better at looking good at fingerprinting than actually being good at fingerprinting (as best i can tell).

That being said, if folks are seeing otherwise, please let me know.  We're not trying to get complacent here :)

<img width="1497" alt="Screen Shot 2021-04-13 at 11 08 05 AM" src="https://user-images.githubusercontent.com/76526/114576386-b2690880-9c40-11eb-9e52-2634f4a14a77.png">

**Comment by TheCultLeader666** [comment]
I don't think that they're just simply using the cache, as my cache clearing extensions do nothing, only when I manually clear the browser cache via settings does my ID reset.

**Comment by pes10k** [comment]
I'm closing this issue since I think we're all resolved that its a mix of cache probe attacks (which should be fixed shortly, with the NetworkIsolationKey work from upstream), the site being clever and trying to claim boring old 1p storage based reidentification is actually fingerprinting prowess, and an agreement that https://github.com/brave/brave-browser/issues/15018 would be a good feature for this issue, and many others.

If folks still have concerns that there are popular scripts successfully reidentifying Brave users through browser fingerprinting, and believes those methods wouldn't be addressed by the further defenses planned in https://github.com/brave/brave-browser/issues/11770, please reopen the issue.

Thanks to everyone who participated in the helpful discussion!
